File: Model_Selection_Boosting/K-Fold_Cross_Validation/K-Fold_Cross_Validation.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Feb 10 10:14:35 2025

@author: swarnabhaghosh
"""

# K-Fold Cross Validation

import os
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Automatically detect the script's directory and set working directory
script_dir = os.path.dirname(os.path.abspath(__file__))
os.chdir(script_dir)

# Importing the dataset
file_path = "Social_Network_Ads.csv"  # Ensure this file is in the same directory as the script

try:
    if os.path.isfile(file_path):
        dataset = pd.read_csv(file_path)
        print("✅ Data imported successfully!")
        X = dataset.iloc[:, [0, 1]].values  # Selecting Age & Salary as features
        y = dataset.iloc[:, 2].values  # Target variable
    else:
        print(f"❌ File not found: {file_path}")
except FileNotFoundError:
    print(f"❌ Error: File not found: {file_path}")

# Splitting dataset into Training & Test sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=0)

# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

# Training the Kernel SVM model on the Training set
from sklearn.svm import SVC
classifier = SVC(kernel='rbf', random_state=0)
classifier.fit(X_train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
accuracy = accuracy_score(y_test, y_pred)

print("\n📌 Confusion Matrix:\n", cm)
print(f"✅ Model Accuracy: {accuracy:.4f}")

# Applying K-Fold Cross Validation
from sklearn.model_selection import cross_val_score
accuracies = cross_val_score(estimator=classifier, X=X_train, y=y_train, cv=10)
print(f"\n📌 K-Fold Accuracy Mean: {accuracies.mean():.4f}")
print(f"📌 K-Fold Accuracy Std Dev: {accuracies.std():.4f}")

# Visualizing Training set results
from matplotlib.colors import ListedColormap
X_set, y_set = X_train, y_train
X1, X2 = np.meshgrid(np.arange(start=X_set[:, 0].min() - 1, stop=X_set[:, 0].max() + 1, step=0.01),
                     np.arange(start=X_set[:, 1].min() - 1, stop=X_set[:, 1].max() + 1, step=0.01))

plt.contourf(X1, X2, classifier.predict(sc.transform(np.array([X1.ravel(), X2.ravel()]).T)).reshape(X1.shape),
             alpha=0.75, cmap=ListedColormap(('red', 'green')))
plt.xlim(X1.min(), X1.max())
plt.ylim(X2.min(), X2.max())

for i, j in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j, 0], X_set[y_set == j, 1], c=ListedColormap(('red', 'green'))(i), label=j)

plt.title('Kernel SVM (Training set)')
plt.xlabel('Age')
plt.ylabel('Estimated Salary')
plt.legend()
plt.show()

# Visualizing Test set results
X_set, y_set = X_test, y_test
X1, X2 = np.meshgrid(np.arange(start=X_set[:, 0].min() - 1, stop=X_set[:, 0].max() + 1, step=0.01),
                     np.arange(start=X_set[:, 1].min() - 1, stop=X_set[:, 1].max() + 1, step=0.01))

plt.contourf(X1, X2, classifier.predict(sc.transform(np.array([X1.ravel(), X2.ravel()]).T)).reshape(X1.shape),
             alpha=0.75, cmap=ListedColormap(('red', 'green')))
plt.xlim(X1.min(), X1.max())
plt.ylim(X2.min(), X2.max())

for i, j in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j, 0], X_set[y_set == j, 1], c=ListedColormap(('red', 'green'))(i), label=j)

plt.title('Kernel SVM (Test set)')
plt.xlabel('Age')
plt.ylabel('Estimated Salary')
plt.legend()
plt.show()

File: Model_Selection_Boosting/K-Fold_Cross_Validation/K-Fold_Cross_Validation.R
# Kernel SVM Classification

# Load Required Libraries
library(caTools)
library(e1071)
library(caret)
library(ggplot2)

# Import Dataset
dataset <- read.csv('Social_Network_Ads.csv')
dataset <- dataset[, 3:5]

# Encoding Target Variable as Factor
dataset$Purchased <- factor(dataset$Purchased, levels = c(0, 1))

# Splitting Dataset into Training and Test Sets
set.seed(123)
split <- sample.split(dataset$Purchased, SplitRatio = 0.75)
training_set <- subset(dataset, split == TRUE)
test_set <- subset(dataset, split == FALSE)

# Feature Scaling
training_set[, -3] <- scale(training_set[, -3])
test_set[, -3] <- scale(test_set[, -3])

# Training Kernel SVM Model
classifier <- svm(formula = Purchased ~ .,
                  data = training_set,
                  type = 'C-classification',
                  kernel = 'radial')

# Predicting Test Set Results
y_pred <- predict(classifier, newdata = test_set[-3])

# Confusion Matrix
cm <- table(test_set[, 3], y_pred)
print(cm)

# Performance Metrics
precision <- cm[2,2] / (cm[2,2] + cm[1,2])
recall <- cm[2,2] / (cm[2,2] + cm[2,1])
f1_score <- 2 * ((precision * recall) / (precision + recall))

print(paste("Precision:", round(precision, 2)))
print(paste("Recall:", round(recall, 2)))
print(paste("F1-Score:", round(f1_score, 2)))

# Applying K-Fold Cross-Validation
set.seed(123)
folds <- createFolds(training_set$Purchased, k = 10)
cv_results <- lapply(folds, function(x) {
  training_fold <- training_set[-x, ]
  test_fold <- training_set[x, ]
  
  classifier <- svm(formula = Purchased ~ .,
                    data = training_fold,
                    type = 'C-classification',
                    kernel = 'radial')
  
  y_pred_fold <- predict(classifier, newdata = test_fold[-3])
  cm_fold <- table(test_fold[, 3], y_pred_fold)
  accuracy_fold <- sum(diag(cm_fold)) / sum(cm_fold)
  return(accuracy_fold)
})
cv_accuracy <- mean(unlist(cv_results))
print(paste("Cross-Validation Accuracy:", round(cv_accuracy * 100, 2), "%"))

# Function for Visualization
visualize_results <- function(set, title, classifier) {
  X1 <- seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.01)
  X2 <- seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.01)
  grid_set <- expand.grid(X1, X2)
  colnames(grid_set) <- c('Age', 'EstimatedSalary')
  
  y_grid <- predict(classifier, newdata = grid_set)
  
  plot(set[, -3],
       main = title,
       xlab = 'Age', ylab = 'Estimated Salary',
       xlim = range(X1), ylim = range(X2))
  contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
  points(grid_set, pch = '.', col = ifelse(y_grid == 1, 'dodgerblue', 'salmon'))
  points(set, pch = 21, bg = ifelse(set[, 3] == 1, 'dodgerblue3', 'salmon3'))
}

# Visualizing Training Set Results
visualize_results(training_set, "Kernel SVM (Training Set)", classifier)

# Visualizing Test Set Results
visualize_results(test_set, "Kernel SVM (Test Set)", classifier)
